### PR TITLE
Optimization of the MaxMin diversity picker

### PR DESCRIPTION
**Pull request by greglandrum**
A contribution from Roger.
The speedup is pretty dramatic.

### CODE REVIEW

File: Code/SimDivPickers/MaxMinPicker.h [hunk 1]
```diff
@@ -133,71 +141,115 @@ RDKit::INT_VECT MaxMinPicker::lazyPick(T &func, unsigned int poolSize,
...
   RDKit::INT_VECT picks;
-  picks.reserve(pickSize);
-  unsigned int pick = 0;
 
-  // enter the pool into a list so that we can pick out of it easily
-  for (unsigned int i = 0; i < poolSize; i++) {
-    pool.push_back(i);
-  }
+  unsigned int memsize = (unsigned int)(poolSize * sizeof(MaxMinPickInfo));
+  MaxMinPickInfo *pinfo = (MaxMinPickInfo *)malloc(memsize);
```
Comment (bp-kelley) [inline Code/SimDivPickers/MaxMinPicker.h]: I suggest using a std::vector here.  malloc has a tendency to lie about allocation anyway.

Comment (bp-kelley) [reply Code/SimDivPickers/MaxMinPicker.h]: or at least new and delete.

File: Code/SimDivPickers/MaxMinPicker.h [hunk 2]
```diff
@@ -133,71 +141,115 @@ RDKit::INT_VECT MaxMinPicker::lazyPick(T &func, unsigned int poolSize,
   if (poolSize < pickSize)
```
Comment (bp-kelley) [inline Code/SimDivPickers/MaxMinPicker.h]: poolSize needs to be > 0, I can't see if that is checked.

Perhaps make both PRECONDITIONS

File: Code/SimDivPickers/MaxMinPicker.h [hunk 3]
```diff
@@ -133,71 +141,115 @@ RDKit::INT_VECT MaxMinPicker::lazyPick(T &func, unsigned int poolSize,
...
-          minTOi = dist;
+    prev = &pool_list;
+    do {
+      poolIdx = *prev;
+      double minTOi = pinfo[poolIdx].dist_bound;
+      if (minTOi > maxOFmin) {
+        unsigned int pi = pinfo[poolIdx].picks;
+        while (pi < picked) {
+          unsigned int pickIdx = picks[pi];
+          CHECK_INVARIANT(poolIdx != pickIdx, "");
```
Comment (bp-kelley) [inline Code/SimDivPickers/MaxMinPicker.h]: Need a readable message for the string I think.

File: Code/SimDivPickers/MaxMinPicker.h [hunk 4]
```diff
@@ -133,71 +141,115 @@ RDKit::INT_VECT MaxMinPicker::lazyPick(T &func, unsigned int poolSize,
...
-        if (dist <= minTOi) {
-          minTOi = dist;
+    prev = &pool_list;
+    do {
+      poolIdx = *prev;
+      double minTOi = pinfo[poolIdx].dist_bound;
+      if (minTOi > maxOFmin) {
+        unsigned int pi = pinfo[poolIdx].picks;
+        while (pi < picked) {
+          unsigned int pickIdx = picks[pi];
```
Comment (bp-kelley) [inline Code/SimDivPickers/MaxMinPicker.h]: pickIdx shadows the one in another scope.

File: Code/SimDivPickers/Wrap/MaxMinPicker.cpp [hunk 5]
```diff
@@ -57,27 +56,14 @@ RDKit::INT_VECT MaxMinPicks(MaxMinPicker *picker, python::object distMat,
...
-    return res;
+    return python::extract<double>(dp_obj(i, j));
   }
 
  private:
   python::object dp_obj;
-  std::map<std::pair<unsigned int, unsigned int>, double> *dp_cache;
 };
 
 RDKit::INT_VECT LazyMaxMinPicks(MaxMinPicker *picker, python::object distFunc,
```
Comment (bp-kelley) [inline Code/SimDivPickers/Wrap/MaxMinPicker.cpp]: Should we add a deprecated message to useCache so we can remove it later?

**Comment by bp-kelley** [comment]
@greglandrum Minor issues, it looks good, I'm happy to test it.  Is there a reference for this method besides Roger's head?

**Comment by greglandrum** [comment]
I'll make those changes.
Done a fair amount of testing already : making sure it generates the same results as the old stuff and that it's faster (which it certainly is).
Roger didn't provide a reference. :-)

File: Code/SimDivPickers/Wrap/MaxMinPicker.cpp (hunk 5 above)
Comment (greglandrum) [reply Code/SimDivPickers/Wrap/MaxMinPicker.cpp]: done

**Pull request merged**